templates/lib/prism/node.rb.erb

```diff
@@ -96,6 +96,39 @@ module Prism
     def type
       raise NoMethodError, "undefined method `type' for #{inspect}"
     end
+
+    # Value equality.
+    def ==(other)
+      return false unless other.is_a?(Node) && other.type == type
+
+      deconstructed_self = deconstruct_keys(nil)
+      deconstructed_other = other.deconstruct_keys(nil)
+
+      return false unless deconstructed_self.keys == deconstructed_other.keys
+
+      deconstructed_self.all? do |key, value|
+        value == deconstructed_other[key]
+      end
+    end
+
+    alias eql? ==
+
+    # Case equality. Implemented like #==, but ignores locations.
+    def ===(other)
+      return false unless other.is_a?(Node) && other.type == type
+
+      deconstructed_self = deconstruct_keys(nil)
+      deconstructed_other = other.deconstruct_keys(nil)
+
+      deconstructed_self.delete(:location)
```

Comment:
`:location` is not the only key that includes locations, there can be others based on `config.yml`.

Overall I feel strongly that we should template this in `node.rb.erb`. As I mentioned above, I don't want to tie the pattern matching implementation to equality.

Reply:
If I understand you correctly, by "template it" you mean implement it in the templated subclasses of `Node`? What would be the advantage of doing that?

Reply:
Notably it would be a lot faster by avoiding array allocations, and also more explicit what it compares (when looking at the generated `node.rb`)

Reply:
> more explicit what it compares

I'm not sure I understand that part. Node equality seems to me like something that should be implemented once for all nodes.

The fact that some node subclasses have additional things that act like fields (`value`) makes that a little difficult, but I think the mixin in `node_ext` is an OK workaround.

On that note. If I were to implement equality in the templated subclasses, I'd have access to `fields` - but I still wouldn't know that for some types I need to compare `value` too. I don't think we want branching on `node.name` in the templating code. The value method is defined in `node_ext`, so that's probably where we should handle it for equality too. Meaning, I think I'd keep `EqualityWithValueMethod` around.

So I could do this two step thing:

1. move the equality methods to the templated subclasses
2. compare all fields from `node.fields` instead of iterating over `deconstruct_keys`

Or alternatively, we add a `fields` class method to each subclass, in which case we could still have only one implementation of equality in the parent class.

I'm struggling a bit to judge what's a good fit here since the templating seems to mean that standard OOP reasoning doesn't apply (otherwise I would strongly argue for just one implementation of node equality)

I'll put up another version and we can see what that looks like.

Reply:
I don't think it's necessary to compare `value` for IntegerNode, FloatNode, etc, *if* locations are compared too.
These `value` methods return derived values from other fields, and given things like `Kernel#Integer` are deterministic, if the other fields (including location) are equal, then the result `#value` must be equal too (because same location and `Integer(str) == Integer(str)`).

But for `===` indeed it's not so simple as that wouldn't compare locations.
I think prepending a module to do `def ===(o); super && value == o.value; end` would be a good way.
An alternative would be some predicate in template.rb.
WDYT @kddnewton ?

Reply:
Interesting points.

Reply:
I'd assume we want `Prism.parse("1") != Prism.parse("2")` though, and since locations are the same here and there is no source file, I think we need to compare the value?